Heat map arrays with indicator overlays

HanKruiger opened the pull request.

HanKruiger commented:
This can be merged now, if it's up to me. I still need to add the color/axis legend to this, but this can be done quickly later.

mbudiu-vmw commented on web/src/main/webapp/rpc.ts:
```diff
@@ -333,3 +333,32 @@ export class ZipReceiver extends Renderer<string> {
+
+// This collector waits for an operation to finish and calls a callback when
+// the operation has completed.
+export class CallbackCollector<T> extends Renderer<T> {
```
I like this.
Maybe it's better to make this a base class which is extended by various renderers.
Let's call it something like OnCompleteRenderer: a renderer which only executes when everything is completed.

mbudiu-vmw commented on web/src/main/webapp/table.ts:
```diff
@@ -41,6 +44,14 @@ import {Dialog} from "./dialog";
+}
+
+export function isCategorical(kind: ContentsKind): boolean {
+    return kind == "Category" || kind == "String"
```
This is wrong, strings are not categories.
Maybe you can call this isString or something else. But I suspect that you don't really want this function.

mbudiu-vmw commented on web/src/main/webapp/table.ts:
```diff
@@ -765,7 +778,19 @@ export class TableView extends RemoteObjectView
 
+    private heatMapArray(): void {
+        let selectedColumns: string[] = [];
+        this.selectedColumns.forEach((col) => selectedColumns.push(col));
```
apparently there are simpler ways to clone an array in javascript.
Maybe you want to add a util.ts function that clones an array?

HanKruiger replied:
The problem here is that `this.selectedColumns` is a set, not an array. But I added functions for both cases in `util.ts` now, so it's simpler.

mbudiu-vmw commented on web/src/main/webapp/heatMapArray/types.ts:
```diff
@@ -0,0 +1,17 @@
+import {IColumnDescription} from "../table"
```
Please note that you have to add some js files to .gitignore when you create new .ts files.
Do you really need a separate folder for this? It only seems to have 2 files.

HanKruiger replied:
I felt like the `webapp` folder was becoming a bit bloated, so I figured to make a folder for the code specific for the heatmap array. But I suppose a separation in terms of functionality (i.e., rpc stuff, ui stuff, etc.) is better than a separation in terms of user features (i.e., heatmap3d, pca, histograms, etc.)? Or do you want to just keep everything in the `webapp` folder?

mbudiu-vmw replied:
This code will further be cleaned-up as it evolves. I have no strict preference on folders yet.

mbudiu-vmw commented on web/src/main/webapp/heatMapArray/view.ts:
```diff
@@ -0,0 +1,288 @@
+import {HeatMapArrayArgs, HeatMapArrayData} from "./types";
+import d3 = require('d3');
+
+class ColorMap {
```
Maybe we can reuse this for the 2d heatmap too. Do you want to refactor that code to use it?

mbudiu-vmw commented on web/src/main/webapp/heatMapArray/view.ts:
```diff
@@ -0,0 +1,288 @@
+
+class CompactHeatMapView implements IHtmlElement {
+    // Size of a single heatmap in pixels
+    public static readonly size: Size = {
```
I am worried about this being hardwired.
First of all, maybe you can reuse the Resolution class.
Second, you should try to compute the size; e.g., if there are only 3 categories, you should make each heat-map 1/3 of the page width instead of hardwiring the size here.

mbudiu-vmw commented on web/src/main/webapp/heatMapArray/view.ts:
```diff
@@ -0,0 +1,288 @@
+
+    public put(x, y, val) {
+        this.chart.append("rect")
+            .attr("x", x * Resolution.minDotSize)
```
You cannot hardwire the dotsize, it has to be computed as the width divided by the number of buckets.
This changes when you resize the window.

mbudiu-vmw commented on web/src/main/webapp/heatMapArray/view.ts:
```diff
@@ -0,0 +1,288 @@
+
+    constructor(private binLabel: string) {
+        this.topLevel = document.createElement("div")
+        if (binLabel.length > CompactHeatMapView.maxTextLabelLength) {
```
Please add a utility function to truncate text.

mbudiu-vmw commented on web/src/main/webapp/heatMapArray/view.ts:
```diff
@@ -0,0 +1,288 @@
+        if (isNumeric(args.cds[0].kind) && isNumeric(args.cds[1].kind) && isCategorical(args.cds[2].kind))
+            this.initiateNumNumCat();
+        else if (isNumeric(args.cds[0].kind) && isNumeric(args.cds[1].kind) && isNumeric(args.cds[2].kind))
+            this.initiateNumNumNum();
```
else throw

mbudiu-vmw commented on web/src/main/webapp/heatMapArray/view.ts:
```diff
@@ -0,0 +1,288 @@
+    }
+
+    public refresh() {
+
```
Please implement this function; it is needed when the browser page is resized.

mbudiu-vmw commented on web/src/main/webapp/heatMapArray/view.ts:
```diff
@@ -0,0 +1,288 @@
+        let oldOffset = this.offset;
+        this.offset = Math.min(
+            Math.floor(position * this.args.uniqueStrings.length),
+            this.args.uniqueStrings.length - this.maxNumHeatMaps()
```
Please create a function totalHeatMaps which just returns args.uniqueStrings.length.

mbudiu-vmw commented on web/src/main/webapp/heatMapArray/view.ts:
```diff
@@ -0,0 +1,288 @@
+    }
+
+    public updateView(heatMapsArray: HeatMapArrayData) {
+        // Clean up before updating.
```
Can you check whether data == null first?

mbudiu-vmw commented on web/src/main/webapp/heatMapArray/view.ts:
```diff
@@ -0,0 +1,288 @@
+    // Entry point. Start the appropriate sketches for the given arguments.
+    public start(args: HeatMapArrayArgs) {
+        this.args = args;
+        if (isNumeric(args.cds[0].kind) && isNumeric(args.cds[1].kind) && isCategorical(args.cds[2].kind))
```
This code is duplicated. Can you factor it out?

mbudiu-vmw commented on web/src/main/webapp/heatmap3d.ts:
```diff
@@ -0,0 +1,804 @@
+        width: 200,
+        height: 200
+    };
+    private static readonly axesMargin = {
```
Can you instead reuse Resolution? (Assuming that the margins can be the same as there).

mbudiu-vmw commented on web/src/main/webapp/heatmap3d.ts:
```diff
@@ -0,0 +1,804 @@
+        top: 40, right: 20, bottom: 40, left: 55
+    };
+
+    private axesG: any;
```
For the `any` fields maybe you can add a comment indicating what they are supposed to be.

mbudiu-vmw commented on web/src/main/webapp/heatmap3d.ts:
```diff
@@ -0,0 +1,804 @@
+    private yText: any;
+
+    // Fields for color legend
+    private static readonly colorLegendSize: Size = {
```
Can you reuse Resolution?

mbudiu-vmw commented on web/src/main/webapp/heatmap3d.ts:
```diff
@@ -0,0 +1,804 @@
+            .attr("y", HeatMapLegend.colorLegendSize.height + HeatMapLegend.colorLegendMargin.bottom - 5);
+    }
+
+    public update(mouseX, mouseY, val) {
```
please add types to the arguments.

mbudiu-vmw commented on web/src/main/webapp/heatmap3d.ts:
```diff
@@ -0,0 +1,804 @@
+        }
+    }
+
+    static invert(v: number, scale: AnyScale, kind: ContentsKind, buckets: string[]): string {
```
This code seems copy-and-paste, maybe we can refactor it in a common class?

mbudiu-vmw commented on web/src/main/webapp/heatmap3d.ts:
```diff
@@ -0,0 +1,804 @@
+    }
+}
+
+export class AxisData {
```
Can you merge this and HeatMap.ts AxisData into one class?

mbudiu-vmw commented on web/src/main/webapp/heatmap3d.ts:
```diff
@@ -0,0 +1,804 @@
+export class HeatMap3DView extends RemoteObjectView implements IScrollTarget {
+    protected topLevel: HTMLElement;
+    public static readonly maxColumns = 4;
+    public static readonly chartWidth = 200;  // pixels
```
Please reuse Resolution.

mbudiu-vmw commented on web/src/main/webapp/heatmap3d.ts:
```diff
@@ -0,0 +1,804 @@
+
+        let rr = this.createRpcRequest("zip", r.remoteObjectId);
+        let renderer = (page: FullPage, operation: ICancellable) => {
+            return new Make2DHistogram(
```
Is this correct or do you want a Make3DHistogram class?

mbudiu-vmw commented on web/src/main/webapp/heatmap3d.ts:
```diff
@@ -0,0 +1,804 @@
+        return max > 20;
+    }
+
+    public static scale(max: number, logScale: boolean) {
```
This stuff is also in common with HeatMapView, maybe you can refactor it into a common base class
HeatMapViewBase

mbudiu-vmw commented on web/src/main/webapp/heatmap3d.ts:
```diff
@@ -0,0 +1,804 @@
+        return this.topLevel;
+    }
+
+    // Generates a string that encodes a call to the SVG translate method
```
Use the function from ui.ts

mbudiu-vmw commented on web/src/main/webapp/heatmap3d.ts:
```diff
@@ -0,0 +1,804 @@
+        rr.invoke(new TableRenderer(page, table, rr, false, order));
+    }
+
+    getHTMLRepresentation(): HTMLElement {
```
this is already inherited from RemoteObjectView, you don't need this method.

mbudiu-vmw commented on web/src/main/webapp/heatmap3d.ts:
```diff
@@ -0,0 +1,804 @@
+    }
+
+    public pageDown() {
+        console.log("TODO: Implement pageDown");
```
I am confused, I thought that I saw code for making pageDown work. Is this needed?

HanKruiger commented:
In the latest commits I've addressed most of your comments. Except for:

* Using the ColorMap in other views. I could do this in a different PR, if I find the time.
* The chart size that we aim for for the heat maps is still hard-coded. But I did make it responsive in the sense that the maximum number of heat maps visible is a function of the page width. Try it out if you want, it looks quite natural.

mbudiu-vmw commented:
I will merge this separately and push the merged version to master.
Then I will close this PR.
